### ISSUE TITLE
FileContentStore: Error on cascade-delete empty directories

### ISSUE THREAD

**Issue by canapa** [Acosix/alfresco-simple-content-stores#22]
Environment:

  - Alfresco Community Edition 6.0 (201806 GA)
  - Module version 1.0.0-SNAPSHOT from master branch

Configuration:

```
simpleContentStores.enabled=true
simpleContentStores.customStores=klSiteRoutingStore,klEncryptedFileStore,klCustomFileStore,defaultTenantFileContentStore
simpleContentStores.rootStore=klSiteRoutingStore

simpleContentStores.customStore.klCustomFileStore.type=standardFileStore
simpleContentStores.customStore.klCustomFileStore.value.rootDirectory=${dir.root}/klContentStore
simpleContentStores.customStore.klCustomFileStore.value.deleteEmptyDirs=true
simpleContentStores.customStore.klCustomFileStore.value.fixedLimit=1073741824

simpleContentStores.customStore.klEncryptedFileStore.type=encryptingFacadeStore
simpleContentStores.customStore.klEncryptedFileStore.ref.backingStore=klCustomFileStore
simpleContentStores.customStore.klEncryptedFileStore.value.keyStorePath=${dir.root}/keystore/klAlfresco.jks
simpleContentStores.customStore.klEncryptedFileStore.value.keyStorePassword=password
simpleContentStores.customStore.klEncryptedFileStore.value.masterKeyAlias=alfcontentstorekey
simpleContentStores.customStore.klEncryptedFileStore.value.masterKeyPassword=password
simpleContentStores.customStore.klEncryptedFileStore.value.keyAlgorithm=RSA
simpleContentStores.customStore.klEncryptedFileStore.value.masterKeyStoreId=klContentStoreKeyId

simpleContentStores.customStore.klSiteRoutingStore.type=siteRoutingStore
simpleContentStores.customStore.klSiteRoutingStore.ref.fallbackStore=defaultTenantFileContentStore
simpleContentStores.customStore.klSiteRoutingStore.map.storeBySitePreset.ref.encrypted-site=klEncryptedFileStore
```

When I upload a pdf document into an encrypted site, I get and error: "Unexpected error occurred during upload of new content".

I found this error in the log file:

```
2019-07-18 17:55:28,628  WARN  [store.file.FileContentStore] [http-apr-8080-exec-1] Error deleting empty parent directories
 java.nio.file.NoSuchFileException: /opt/alfresco/alf_data/contentstore/2019/7/18/17/55
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.isSameFile(UnixFileSystemProvider.java:338)
	at java.nio.file.Files.isSameFile(Files.java:1504)
	at de.acosix.alfresco.simplecontentstores.repo.store.file.FileContentStore.deleteEmptyParents(FileContentStore.java:614)
	at de.acosix.alfresco.simplecontentstores.repo.store.file.FileContentStore.delete(FileContentStore.java:423)
	at org.alfresco.repo.content.AbstractRoutingContentStore.delete(AbstractRoutingContentStore.java:407)
	at de.acosix.alfresco.simplecontentstores.repo.store.routing.MoveCapableCommonRoutingContentStore.delete(MoveCapableCommonRoutingContentStore.java:391)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:338)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:197)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at de.acosix.alfresco.simplecontentstores.repo.aop.InitContentStoreContextInterceptor.lambda$invoke$0(InitContentStoreContextInterceptor.java:76)
	at de.acosix.alfresco.simplecontentstores.repo.store.context.ContentStoreContext.executeInNewContext(ContentStoreContext.java:143)
	at de.acosix.alfresco.simplecontentstores.repo.aop.InitContentStoreContextInterceptor.invoke(InitContentStoreContextInterceptor.java:58)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at de.acosix.alfresco.simplecontentstores.repo.aop.ContentStoreCapsEmulatingInterceptor.invoke(ContentStoreCapsEmulatingInterceptor.java:194)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy26.delete(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.alfresco.repo.management.subsystems.SubsystemProxyFactory$1.invoke(SubsystemProxyFactory.java:79)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.alfresco.repo.management.subsystems.CryptodocSubsystemProxyFactory$1.invoke(CryptodocSubsystemProxyFactory.java:60)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
	at com.sun.proxy.$Proxy26.delete(Unknown Source)
	at org.alfresco.repo.content.cleanup.EagerContentStoreCleaner.deleteFromStore(EagerContentStoreCleaner.java:325)
	at org.alfresco.repo.content.cleanup.EagerContentStoreCleaner.deleteFromStores(EagerContentStoreCleaner.java:308)
	at org.alfresco.repo.content.cleanup.EagerContentStoreCleaner.afterRollback(EagerContentStoreCleaner.java:254)
	at org.alfresco.util.transaction.TransactionSupportUtil$TransactionSynchronizationImpl.afterCompletion(TransactionSupportUtil.java:603)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.invokeAfterCompletion(TransactionSynchronizationUtils.java:171)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.invokeAfterCompletion(AbstractPlatformTransactionManager.java:992)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.triggerAfterCompletion(AbstractPlatformTransactionManager.java:967)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processRollback(AbstractPlatformTransactionManager.java:869)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.rollback(AbstractPlatformTransactionManager.java:812)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.completeTransactionAfterThrowing(TransactionAspectSupport.java:550)
	at org.alfresco.util.transaction.SpringAwareUserTransaction.completeTransactionAfterThrowing(SpringAwareUserTransaction.java:574)
	at org.alfresco.util.transaction.SpringAwareUserTransaction.rollback(SpringAwareUserTransaction.java:544)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:519)
	at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecute(RepositoryContainer.java:587)
	at org.alfresco.repo.web.scripts.RepositoryContainer.transactionedExecuteAs(RepositoryContainer.java:656)
	at org.alfresco.repo.web.scripts.RepositoryContainer.executeScriptInternal(RepositoryContainer.java:428)
	at org.alfresco.repo.web.scripts.RepositoryContainer.executeScript(RepositoryContainer.java:308)
	at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:399)
	at org.springframework.extensions.webscripts.AbstractRuntime.executeScript(AbstractRuntime.java:210)
	at org.springframework.extensions.webscripts.servlet.WebScriptServlet.service(WebScriptServlet.java:132)

```

**Comment by AFaust** [comment]
As mentioned in #21 I hope to have some time this weekend to look at this.

**Comment by AFaust** [comment]
In my test, this kind of error was the result of a preceding error. The masterKeyAlias "alfcontentstorekey" is too long for the column in the Alfresco database. An alias can at most be 15 characters long. If the alias was too long, I got that error. Anyway, I have addressed this issue and made the deletion safer, so that such a warning does not further confuse a user / admin. I have applied the patch to master and alfresco-5.0 branches, but I still have to apply it to the alfresco-4.2 branch, which I still have to update to the latest state of my Maven framework / utility module (other branches have already been adapted today).
